service/frontend/nexus_incoming_service_client.go

```diff
@@ -0,0 +1,373 @@
+			return nil, serviceerror.NewInternal("error listing Nexus incoming services")
+		}
+
+		for _, entry := range resp.Entries {
+			if request.Name == entry.Service.Spec.Name {
+				result.Services = []*nexus.IncomingService{persistedEntryToExternalAPI(entry)}
+				return result, nil
+			}
+		}
+
+		if resp.NextPageToken == nil {
+			return result, nil
+		}
+
+		currentPageToken = resp.NextPageToken
+	}
+
+	return nil, ctx.Err()
+}
+
+func persistedEntryToExternalAPI(entry *persistencespb.NexusIncomingServiceEntry) *nexus.IncomingService {
+	var lastModifiedTime *timestamppb.Timestamp
+	// Only set last modified if there were modifications as stated in the UI contract.
+	if entry.Version > 1 {
+		lastModifiedTime = timestamppb.New(hlc.UTC(entry.Service.Clock))
+	}
+
+	return &nexus.IncomingService{
+		Version:          entry.Version,
+		Id:               entry.Id,
+		Spec:             entry.Service.Spec,
+		CreatedTime:      entry.Service.CreatedTime,
+		LastModifiedTime: lastModifiedTime,
+		UrlPrefix:        "/" + cnexus.Routes().DispatchNexusTaskByService.Path(entry.Id),
+	}
+}
+
+func (c *NexusIncomingServiceClient) getServiceNameIssues(name string) rpc.RequestIssues {
+	var issues rpc.RequestIssues
+
+	if name == "" {
+		issues.Append("incoming service name not set")
+		return issues
+	}
+
+	maxNameLength := c.config.maxNameLength()
+	if len(name) > maxNameLength {
+		issues.Appendf("incoming service name exceeds length limit of %d", maxNameLength)
+	}
+
+	if !ServiceNameRegex.MatchString(name) {
+		issues.Appendf("incoming service name must match the regex: %q", ServiceNameRegex.String())
+	}
+
+	return issues
+}
+
+func (c *NexusIncomingServiceClient) validateUpsertSpec(spec *nexus.IncomingServiceSpec) error {
+	issues := c.getServiceNameIssues(spec.GetName())
+
+	if spec.Namespace == "" {
+		issues.Append("incoming service namespace not set")
+	} else if _, nsErr := c.namespaceRegistry.GetNamespace(namespace.Name(spec.Namespace)); nsErr != nil {
+		return serviceerror.NewFailedPrecondition(fmt.Sprintf("could not verify namespace referenced by incoming service exists: %v", nsErr.Error()))
+	}
+
+	if spec.TaskQueue == "" {
+		issues.Append("incoming service task queue not set")
+	} else {
+		maxTaskQueueLength := c.config.maxTaskQueueLength()
+		if len(spec.TaskQueue) > maxTaskQueueLength {
+			issues.Appendf("incoming service task queue exceeds length limit of %d", maxTaskQueueLength)
+		}
+		if strings.HasPrefix(spec.TaskQueue, reservedTaskQueuePrefix) {
+			issues.Appendf("incoming service task queue begins with reserved prefix: %q", reservedTaskQueuePrefix)
+		}
+	}
+
+	maxSize := c.config.maxSize()
+	if spec.Size() > maxSize {
+		issues.Appendf("incoming service size exceeds limit of %d", maxSize)
+	}
+
+	return issues.GetError()
+}
+
+func getServiceIdIssues(ID string) rpc.RequestIssues {
+	var issues rpc.RequestIssues
+	if ID == "" {
+		issues.Append("incoming service Id not set")
```

Comment:
nit: use `ID` in Go code and English. The fact that the proto generated Go code produced a field called `Id` shouldn't affect our error messages since in other languages the field might be in lower case.